src/NuGet.Core/NuGet.Packaging/ContentModel/ManagedCodeConventions.cs

```diff
@@ -22,7 +21,7 @@ public class ManagedCodeConventions
 
         private static readonly ContentPropertyDefinition AnyProperty = new ContentPropertyDefinition(
             PropertyNames.AnyValue,
-            parser: (o, t) => o); // Identity parser, all strings are valid for any
+            parser: (o, t) => o.ToString()); // Identity parser, all strings are valid for any
```

Comment:
When does this run? I would be worried that this defeats the purpose of the change.

Reply:
I am not an expert in this code path, but my understanding is that any pattern that has `{any}` or 
`{any?}` path segment will execute this code path. Looking at the code https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/ContentModel/ManagedCodeConventions.cs#L421-L573, almost all patterns except MSBuild related files have the said `any` path segment in them. I am also worried that it might defeat the purpose of the change. However, a PerfView trace before and after the fix would help.

Reply:
Yeah this is for path segments that match on everything for example `lib/{tfm}/{any}`.  That case, the `o.ToString()` makes a string out of something like `PackageA.dll` which gets added to the properties in `TokenSegment.TryMatch()`.  This means the substring was used so its okay to realize it to a string?  Since the point was to stop allocating a substring if its not used.

Reply:
> Since the point was to stop allocating a substring if its not used.

I agree with this goal. Is there any test that checks for the scenario where substrings are not used? The reason is that the pattern matching algorithm is executed during the pack and restore operation. Hence, I am curious to understand the scenario where substrings are not used. For example, I debugged first 3-unit tests in https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Packaging.Test/ContentModelTests/ContentModelLibTests.cs and noticed that substrings are invoked every time. Maybe I chose wrong tests to debug.

Reply:
@kartheekp-ms Are you saying are created for _every_ segment in those strings?

Reply:
Substrings are created if the segment satisfies a particular parser logic. For example, if it is a `{any}` segment and `path` has `xyz.dll`, the code creates a substring. Likewise looking at all the [parsers logic](https://github.com/NuGet/NuGet.Client/blob/dev-jeffkl-tokensegment-trymatch/src/NuGet.Core/NuGet.Packaging/ContentModel/ManagedCodeConventions.cs#L19-L37), there is `.ToString()` invocation which is invoked if the token satisfies a path segment.

Reply:
A well authored package will match some sort of a pattern we have.

Example, you won't see packages with a `bla/something.txt` often, whih is the only place it won't get actualized. 

I think it's a big reason why we're not seeing the improvements we're hoping for.

Reply:
A more performant approach would probably be a way to figure out a way to deduplicate strings or rewrite the matching. 

That can be done, but tests are showing that this change is an improvement as well, so we can get it merge and work on other improvements separately.